Custom GrowthRatios and more Schedule variables

DavidAnderegg opened the pull request:
## Purpose
~~This PR adds the ability to set custom growth ratios on each layer. Additionally, more 'schedule' variables are introduced.~~

This PR adjusts how some options work. Apart from prescribing a single value, it is now possible to define a list with per-layer values. Additionally, it is possible to instruct pyHyp to interpolate the value linearly throughout the extrusion.

This PR removes the `volSmoothSchedule` variable as the same functionality is now possible with  `volSmoothIter`.

Lastly, a new variable `growthRatios` is introduced to explicitly set the growth ratio. If it is set, the `marchDist` option is ignored. This variable works like the other changes introduce here. 


## Expected time until merged
1 week

## Type of change
<!--
What types of change is it?
Select the appropriate type(s) that describe this PR
-->

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (non-backwards-compatible fix or feature) - `volSmoothSchedule` is now obsolete and was removed
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no API changes)
- [ ] Documentation update
- [ ] Maintenance update
- [ ] Other (please describe)

## Testing
I adjusted the `naca0012_rans` and `corner` test to include an 'interpolation' and a 'per-layer' case. I also added a test for `nConstantStart` and `nConstantEnd` options. And finally, i added a test for `simpleOCart` with `xBounds` and no nearfield mesh.

~~Here are the new ref-files for the new tests: [schedule_ref_files.zip](https://github.com/user-attachments/files/16871720/schedule_ref_files.zip)~~

Here are the new ref-files for the new tests: [refFiles.zip](https://github.com/user-attachments/files/17148112/refFiles.zip)



<!-- Explain the steps needed to test the new code to verify that it does indeed address the issue and produce the expected behavior. -->

## Checklist
<!-- Put an `x` in the boxes that apply. -->

- [x] I have run `flake8` and `black` to make sure the Python code adheres to PEP-8 and is consistently formatted
- [x] I have formatted the Fortran code with `fprettify` or C/C++ code with `clang-format` as applicable
- [x] I have run unit and regression tests which pass locally with my changes
- [x] I have added new tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation

eirikurj commented:
@DavidAnderegg thanks for the PR. Will try to take a look as soon as I can. In the meantime, can you address the failing tests.

DavidAnderegg commented:
@eirikurj The tests fail because new the ref files have not been added. I attached a .zip in the description of this PR.

eirikurj commented:
My bad, I only read the description in a rush and saw the test are failing. I can update the reference files. I skimmed the PR content, and it looks good, but will do a more thorough pass later. 
I think it would be good if we can combine or refactor the RANS scripts to avoid code duplication. They are essentially all the same except for a few options. I suggest one of the two,
1. refactor the geometry generation out into its own file that can be imported or run. 
2. use `argparse` with argument `mode` (or similar) to run each case and just update the relevant options based on the mode. This will have a smaller footprint.

eirikurj commented:
I have updated the files. However, in line with my previous comment on refactoring the scripts I updated the names and inserted `rans` in the files i.e.,
```
naca0012_growth_ratios.cgns -> naca0012_rans_growth_ratios.cgns
naca0012_schedule.cgns      -> naca0012_rans_schedule.cgns
```

DavidAnderegg commented:
I agree, getting rid of the duplicated code is a good thing. I adjusted that and also fixed a tiny bug where the marchDistance is not computed when using custom growth ratios. I think this bug is more a testing issue because the test uses the marchDistance to compare the files.

Also, the ref file for the growth ratios needs to be updated: [growth_ratios_ref.zip](https://github.com/user-attachments/files/16884815/growth_ratios_ref.zip)

eirikurj commented on src/3D/3D_utilities.F90:
```diff
@@ -58,14 +58,15 @@ subroutine calcGridRatio(N, nStart, nEnd, s0, S, ratio)
     integer(kind=intType), intent(in) :: N, nStart, nEnd
-    real(kind=realType), intent(in) :: s0, S
+    real(kind=realType), intent(in) :: s0
+    real(kind=realType), intent(inout) :: S
```
Related to my previous comment. The way that this function `calcGridRatio` is currently used is to pass in the `marchDist`, which is an input parameter. This will then modify that value, which is not great.

eirikurj reviewed (changes requested):
To avoid having this sit too long, I did another quick pass over. Overall looks good, but please take a look at my comments.

DavidAnderegg commented on pyhyp/pyHyp.py:
```diff
@@ -601,16 +602,21 @@ def _getDefaultOptions():
             "epsE": [float, 1.0],
             "epsI": [float, 2.0],
             "theta": [float, 3.0],
+            "epsESchedule": [(list, type(None)), None],
```
I agree.

DavidAnderegg commented:
@eirikurj I thought about your comment regarding the modification of input-variables. I think the problem is that I pass the same variable in multiple versions to the fortran layer. E.g I pass `splay` and `splaySchedule`, which is the same, but the first time it is a scalar and the second time, it is an interpolation instruction. This leads to the need to figure out which one is used in the fortran layer which is the reason to this 'clusterfuck' you rightfully criticize.

I think a better approach would be to pass only one `splay` variable to the fortran layer. But instead of a scalar, this will be a list with the same number of entries as there are layers (similar to how growthRatios is implemented). For this to work, I would need to move the 'interpolate logic' to the python layer. But it would also add the benefit that one could specify 'splay' as a scalar, an interpolation or even an list of values for each layer.

I think this is a better approach. If you dont disagree, this is the path I will follow.

DavidAnderegg commented on pyhyp/utils.py:
```diff
@@ -67,6 +67,11 @@ def simpleOCart(inputGrid, dh, hExtra, nExtra, sym, mgcycle, outFile, userOption
         input_filename = inputGrid
         inputGrid = readGrid(input_filename)
+    # the input Grid can be None, in this acse, we must have xBounds
+    elif inputGrid is None:
```
Added a test.

DavidAnderegg commented on doc/options.yaml:
```diff
@@ -235,6 +235,40 @@ volSmoothSchedule:
+
+growthRatios:
+  desc: >
+    Manually specify the growth ratio for each layer. This must be a list with :py:data:`N` - 1 ratios. 
```
added example

DavidAnderegg commented:
@eirikurj I adjusted the code as explained in my previous comment. It is once again ready for review.

DavidAnderegg commented on doc/options.yaml:
```diff
@@ -94,20 +94,28 @@ splay:
   desc: >
     How hard to try to force orthogonality at splay edges.
     Should be between 0 and 0.5.
+    Please look at :py:data:`splay` for explanations and examples on the format.
```
I also added whatever I thought would make sense. I am open to remove/add even more.

DavidAnderegg commented on pyhyp/pyHyp.py:
```diff
@@ -601,16 +620,16 @@ def _getDefaultOptions():
-            "volCoef": [float, 0.25],
-            "volBlend": [float, 0.0001],
-            "volSmoothIter": [int, 100],
-            "volSmoothSchedule": [(list, type(None)), None],
```
I added a deprecation warning.

DavidAnderegg commented:
@anilyil, I adressed your comments. You might want to take a look again.

DavidAnderegg commented:
@anilyil I know why this happens... Yes, I will look into it.

DavidAnderegg commented:
@anilyil Grid Ratios for pyHypMulti print now correctly.

I also replaced that code with 'tabulate' because it is so much easier this way.

eirikurj commented on pyhyp/pyHyp.py:
```diff
@@ -758,14 +742,204 @@ def _setOptions(self):
+            if pGridRatio > min_growth_ratio:
+                raise Error(f"The `pGridRatio` option has to be lower than the lowest grid ratio ({min_growth_ratio})")
+
+        # figure out initial pseud grid initial offwall spacing
```
typo "pseud"

eirikurj commented on pyhyp/pyHyp.py:
```diff
@@ -778,6 +952,12 @@ def __del__(self):
         self.hyp.releasememory()
 
+    def get_used_march_distance(self):
+        if self.gridGenerated:
+            return self.hyp.hypinput.marchdist
+        else:
+            raise Error("Can not returning used marching distance before extruding the grid")
```
No need for the `else`, just check the inverse i.e., `if not self.gridGenerated` and throw the error, otherwise 
just `return` the member variable.
Also please rephrase the error string.

eirikurj commented on pyhyp/pyHyp.py:
```diff
@@ -758,14 +742,204 @@ def _setOptions(self):
+
+        return fullDeltaS, marchDist, growth_ratios
+
+    def get_growth_ratio_string(self, growth_ratios=None, n_decimals=3):
```
Add argument descriptions (same for other new functions). Also probably should make it "private" with an underscore, or is there a use case for the user here that I am missing?

DavidAnderegg replied:
I think this depends on your definition of private/public. I understood it like this: only accessed within the class = private; gets accessed outside of the class = public.

This function is called by pyHypMulti, which does not inherit from pyHyp. Thus it is accessed from the outside and should be public.

eirikurj replied:
Forgot that its used in `pyHypMulti`, so this is fine as is.

eirikurj reviewed (changes requested):
Sorry for the delay. I did a quick pass and overall the changes are very good. Just a few very minor comments or questions. Let me know if anything is unclear. Also, can you prepare and link an archive with updated reference files for the tests. Want to test this locally and I will also need to upload it to our site.

DavidAnderegg commented on pyhyp/pyHyp.py:
```diff
@@ -758,14 +742,204 @@ def _setOptions(self):
+
+        return growth_ratio_string
+
+    def _configure_pseudo_grid_parameters(self, growth_ratios):
```
There is no need. I just added it as member variable after the fact. 

I will change it.

DavidAnderegg commented on pyhyp/pyHyp.py:
```diff
@@ -25,16 +26,35 @@
 from cgnsutilities.cgnsutilities import readGrid, combineGrids
 
 
+class pyHypWarning(object):
```
I can move to a standard library. I just copied the idea from ADflow. Let me know what you want?

eirikurj replied:
I think this is fine for now. Overall, I would lean towards a more standard approach, but both have merits.

DavidAnderegg commented on pyhyp/pyHyp.py:
```diff
@@ -758,14 +742,204 @@ def _setOptions(self):
+
+        return fullDeltaS, marchDist, growth_ratios
+
+    def get_growth_ratio_string(self, growth_ratios=None, n_decimals=3):
+        if growth_ratios is None:
+            growth_ratios = self.growth_ratios
```
you are correct.

eirikurj commented on pyhyp/pyHyp.py:
```diff
@@ -405,8 +384,19 @@ def __init__(self, comm=None, options=None, debug=False):
             raise Error("The options = keyword argument is *NOT* optional. " "It must always be provided")
 
+        # Deprecated options
+        deprecated_options = self._getDeprecatedOptions()
```
Thanks. Yes, there are some plugins available for `flake8` (such as [pep8-naming](https://github.com/PyCQA/pep8-naming)). Ideally we should enforce this, but this might take some time to enforce since there will be many instances that need to be changed (and some disabled). Will open an issue to discuss (see https://github.com/mdolab/.github/issues/66)

DavidAnderegg replied:
Thanks, sounds good.

DavidAnderegg commented:
@anilyil The new ref files do break the old tests. This is because I extended the 'corner' case. So I suggest you wait with updating until the day this PR gets merged.

DavidAnderegg commented on pyhyp/pyHyp.py:
```diff
@@ -758,14 +742,296 @@ def _setOptions(self):
+    def getGrowthRatioString(self, nDecimals=3):
+        """
+        Returns a rounded string with the current growth ratio. If the growth
+        ratio is constant, a single value is return. Otherwise a range is
```
thanks.

DavidAnderegg commented on pyhyp/pyHyp.py:
```diff
@@ -229,11 +249,17 @@ def __init__(self, comm=None, options=None, commonOptions=None, debug=False, ski
-                self.results["gridRatio"][index] = float(hypGrid.hyp.hypdata.gridratio)
-                self.results["minQualityOverall"][index] = float(hypGrid.hyp.hypdata.minqualityoverall)
-                self.results["minVolumeOverall"][index] = float(hypGrid.hyp.hypdata.minvolumeoverall)
+                self.results["minQualityOverall"][index] = roundSig(
```
Yes, I only round for stdout. I agree, I should probably move the rounding business to the print function and leave the raw number in `self.results`.

I don't use `floatfmt` because it only allows rounding to a certain digit after the decimal point. But my function rounds to significant figures, no matter where the decimal point is. I forgot exactly why I introduced it, but I believe the 'minQualityOverall' would either be a very long string or just 0. I think rounding to significant figures is better.

Let me know what you think.

DavidAnderegg replied:
I moved the rounding into the print function.

eirikurj replied:
Thanks for separating the data and the stdout process. I think I understand this a bit better now, although I think a scientific notation, e.g., `floatfmt=".4e"`, would give you the consistent output needed. I understand it's not as nice and easy to read as the float/decimal format, and may not be desired in some places. 
I think using `roundSig` is probably fine, but might give the user the wrong idea. We should try to present these values carefully and make sure its not rounded too heavily. 4 seems a bit small but this depends on what is being presented. A number like `12345.6789` with 4 significant digits as used here would be rounded to `12350` using `roundSig`, but `.4f` would present the number as is, and using `.4e` as `1.2346e+04`.

I dont have many broad cases to test this on, and I want to get stuck on this as this is a minor thing, but perhaps @anilyil has any input or comments here. If not we can go ahead and merge.

DavidAnderegg replied:
Somehow, the scientific notation did not even come to my mind...

I changed it locally but it does look worse. So not sure, I agree with your point about confusing the user. I am open to both solutions but I lean towards the scientific notation because it reduces ambiguity. 

Lets wait for @anilyil's opinion.

DavidAnderegg replied:
@anilyil agrees with scientific, so lets do this then.

DavidAnderegg commented on pyhyp/pyHyp.py:
```diff
@@ -758,14 +742,296 @@ def _setOptions(self):
+        growthRatioString = self.getGrowthRatioString(nDecimals)
+
+        print("#--------------------#")
+        print(f"Grid Ratio:  {growthRatioString}")
+        print(f"March Dist:  {roundSig(self.marchDistance, nDecimals)}")
```
Same as above, I think significant figures rounding is better. 

Normal formatting either prints a lot of decimals for farfield meshes or not enough for near field meshes.

DavidAnderegg replied:
see above, moving to scientific notation.

DavidAnderegg commented on src/utils/releaseMemory.F90:
```diff
@@ -57,8 +57,40 @@ subroutine releaseMemory
+    if (allocated(splayCornerOrthogonality)) then
+        deallocate (splayCornerOrthogonality)
     end if
 
```
Good catch. Thank you.

DavidAnderegg commented on pyhyp/pyHyp.py:
```diff
@@ -758,14 +742,296 @@ def _setOptions(self):
+        growthRatios : list[float]
+            The growth ratio for each layer.
+        """
+        # initial ratio and r is the grid ratio.
+        # function 'f' is S - s0*(1-r^n)/(1-r) where S is total length, s0 is
```
I am impressed. Thank you.

DavidAnderegg commented:
@eirikurj I am sorry for the missing ref-file, I did not realize it... I addressed the comments that were clear. I also changed `get-ref-files.sh` to point towards `pyhyp_ref_files_20241210.tar.gz`.

Please let me know what you think regarding the not resolved points.

DavidAnderegg commented:
Regarding the failed tests. This is very strange to me. Everything passes except for `test_all_examples.py:TestExamples.testCorner
`. To further add to the confusion, this test also passes on `u22-gcc-ompi-latest` but fails on the other 3 images. I can reproduce this error locally, but I am a bit lost on how to debug it. Any suggestions are welcome...

eirikurj commented on pyhyp/pyHyp.py:
```diff
@@ -887,3 +1145,33 @@ def generateOutputName(inputFile, outputType):
+    growthRatios : list(float)
+        List holding the growth ratio for each layer
+
+    nDecimals : int
+        The number of significant figures to round to.
+
```
Please remove this (outdated)

eirikurj commented:
Looks good, thanks for all your efforts on this!
